### PR TITLE
Add MaxBin 2

### PR DESCRIPTION
**Pull request by colinbrislawn**
* [x] I have read the guidelines above.
* [x] This PR adds a new recipe.
* [ ] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

### CODE REVIEW

**Comment by colinbrislawn** [comment]
When I build this package localy with `conda build .` it fails with this error. 
```
g++ -c -Wall -g -std=c++0x EManager.cpp -o EManager.o -I ./
In file included from EManager.cpp:1:
./EManager.h:12:10: fatal error: 'thread' file not found
#include <thread>
         ^
1 error generated.
make: *** [EManager.o] Error 1
```

When I manually move into the working directory and run `make`, it compiles successfully with some warnings. Strange...

The `#include <thread>` error seems to be related to C++11. Any ideas?

**Comment by colinbrislawn** [comment]
I've noticed that @keuv-grvl has crafted excellent recipes for related projects. Do you have any advice for me on this build?

Thanks!

File: recipes/maxbin2/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    - fraggenescan >=1.30
+    - hmmer
+    - idba
+test:
+  commands:
+    - MaxBin > /dev/null
+about:
+  home: "http://downloads.jbei.org/data/microbial_communities/MaxBin/MaxBin.html"
+  license: "BSD 3-clause"
+  summary: "MaxBin is software for binning assembled metagenomic sequences based on an Expectation-Maximization algorithm."
```
Comment (keuv-grvl) [inline recipes/maxbin2/meta.yaml]: I find useful to add an `extra` category with the recipe maintainer(s) and the associated DOI pointing to the original article (if available). Here is an example: [abawaca](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/abawaca/meta.yaml#L29-L32)

**Comment by keuv-grvl** [comment]
Now I remember why I crafted the FragGeneScan recipe!

`<thread>` is quite new and required GCC 4.6+. Moreover the `makefile` states `LDFLAGS:=-pthread`. According to my c++-enthusiast colleague, it should be `-lpthread`. He helped me rewrite the makefile. You can find it [here](https://gist.github.com/keuv-grvl/92f0e22e769dec533a8e69344777cc7c).

Also you may test your recipes using [`./simulate-travis.py`](https://github.com/bioconda/bioconda-recipes/blob/master/FAQs.md#test-locally) to avoid interfering with your OS.

File: recipes/maxbin2/build.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+set -euo pipefail
+cd src && make
```
Comment (colinbrislawn) [inline recipes/maxbin2/build.sh]: I modeled cd src && make after the docker build file. I like your idea of copying the executable to the right location.

File: recipes/maxbin2/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  run:
+    - libgcc # [linux]
+    - perl
+    - bowtie2
+    - fraggenescan >=1.30
+    - hmmer
+    - idba
+test:
+  commands:
+    - MaxBin > /dev/null
```
Comment (colinbrislawn) [inline recipes/maxbin2/meta.yaml]: I did notice that the return code was 255. Usually I would test for installation using 
`MaxBin --help` or `MaxBin --version`, but the program only accepts real inputs. Any ideas?

File: recipes/maxbin2/meta.yaml (hunk 1 above)
Comment (colinbrislawn) [inline recipes/maxbin2/meta.yaml]: Added!

File: recipes/maxbin2/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  fn: MaxBin-2.2.1.tar.gz
+  url: http://downloads.jbei.org/data/microbial_communities/MaxBin/getfile.php?MaxBin-2.2.1.tar.gz
+  md5: 3748e8f65b5bf18f8bd341210724cc5c
+requirements:
+  build:
+    - gcc # [linux]
+    - llvm # [osx]
+  run:
+    - libgcc # [linux]
+    - perl
```
Comment (colinbrislawn) [inline recipes/maxbin2/meta.yaml]: I don't think my collaborators need perl, but I'll check. Hopefully we can drop this.

**Comment by colinbrislawn** [comment]
Thank you for all your feedback. This is exactly the kind of careful review that this recipe needed. 

I've addressed as many suggestions as possible and added your approach to program versioning. 

What should we do about the makefile? Patch it, or perhaps submit a PR to maxbin?

**Comment by daler** [comment]
@colinbrislawn some build system issues were fixed over in #3080 if you want to rebase and give this another shot.

File: recipes/maxbin2/meta.yaml (hunk 1 above)
Comment (keuv-grvl) [inline recipes/maxbin2/meta.yaml]: :tada:

File: recipes/maxbin2/meta.yaml (hunk 3 above)
Comment (keuv-grvl) [inline recipes/maxbin2/meta.yaml]: I had this problem with #2618 and we solved it by testing if the return code was 255.
Providing reads as input would make a functional test but this is not the bioconda's aim (cf. also #2618)

File: recipes/maxbin2/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+about:
+  home: "http://downloads.jbei.org/data/microbial_communities/MaxBin/MaxBin.html"
+  license: "BSD 3-clause"
+  summary: "MaxBin is software for binning assembled metagenomic sequences based on an Expectation-Maximization algorithm."
+
+extra:
+  recipe-maintainers:
+    - colinbrislawn
+    - keuv-grvl
+  doi: 10.1186/2049-2618-2-26
```
Comment (keuv-grvl) [inline recipes/maxbin2/meta.yaml]: You could also add MaxBin 2.0 paper : [10.1093/bioinformatics/btv638](https://dx.doi.org/10.1093/bioinformatics/btv638)
``` yaml
doi:
  - 10.1186/2049-2618-2-26
  - 10.1093/bioinformatics/btv638
```

File: recipes/maxbin2/build.sh [hunk 6]
```diff
@@ -1,3 +1,4 @@
 #!/bin/bash
 set -euo pipefail
 cd src && make
+cp src/MaxBin $PREFIX/bin/
```
Comment (keuv-grvl) [inline recipes/maxbin2/build.sh]: Your script is already located in `src` at this point so `src/MaxBin` does not exist but `./MaxBin` does

**Comment by colinbrislawn** [comment]
@keuv-grvl  Done! Thanks for all your feedback.

@daler I'll rebase this afternoon. Thank you for supporting this dev platform!

**Comment by keuv-grvl** [comment]
I just try to run MaxBin (manually installed) using the `run_MaxBin.pl` script.

For the complete pipeline, the following files are required in `$PREFIX/bin/` :
- `run_MaxBin.pl`
- `_getabund.pl`
- `_getmarker.pl`
- `_sepReads.pl`
- `heatmap.r`
- `marker.hmm`
- `bacar_marker.hmm`
- `src/MaxBin`

The last one is compile from sources exactly like in your `build.sh`. The `run_MaxBin.pl` search for it in a `src` folder.

So if you plan to use the complete pipeline with `run_MaxBin.pl`, you will need to compile the sources then copy some files.

Hope this helps :)

**Comment by colinbrislawn** [comment]
@keuv-grvl That does help! I'll make sure to copy files to all needed locations.

I've delayed this rebase while working on other projects, but also because I ran into an issue incorporating the `maxbin.makefile` that you and @Celforyon made. How can I get this into the `src` directory so I can use it during the build?

**Comment by keuv-grvl** [comment]
I like to use patches because they are managed by Conda.

First, you build your patch:
``` bash 
cd /path/to/bioconda-recipe/recipes/maxbin/
wget https://gist.githubusercontent.com/keuv-grvl/92f0e22e769dec533a8e69344777cc7c/raw/01719d5e63118deee3bbfa0c75d56c36917f76d7/maxbin.makefile
diff -u /dev/null maxbin.makefile > makefile.patch # this compares an empty file to maxbin.makefile
# you can test your patch as follows:
patch < makefile.patch
```

Then you add it to your `meta.yaml`:
``` 
source:
  fn: ...
  url: ...
  md5: ...
  patches:
    - makefile.patch
```

Patches are applied before calling `build.sh`, so `maxbin.makefile` will exist.
In your `build.sh`:
``` bash
cp maxbin.makefile src/
cd src/
make -f maxbin.makefile
```

**Comment by colinbrislawn** [comment]
Thanks to help from @keuv-grvl and @ajaypanyala!

Once this building, I think we are ready to merge.

**Comment by keuv-grvl** [comment]
Travis failed for OSX [here (L360-L362)](https://travis-ci.org/bioconda/bioconda-recipes/jobs/194634556#L360-L362). It seems to come from the C++11 support. This requires additional parameters (see [here](http://stackoverflow.com/questions/14115314/how-to-compile-c11-with-clang-3-2-on-osx-lion)). But I bet these parameters are OSX specific, so you may need to set `CXX_FLAGS` only for OSX (something like [this](https://github.com/bioconda/bioconda-recipes/blob/013af80cae89e5c403cb101e622d6f5156b30bee/recipes/pbbam/build.sh#L5-L7)).

Otherwise, you can simply skip the build for OSX.

File: recipes/maxbin2/meta.yaml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-{% set name = "maxbin" %}
```
Comment (bgruening) [inline recipes/maxbin2/meta.yaml]: This will change the entire package name, is this intended?

Comment (colinbrislawn) [inline recipes/maxbin2/meta.yaml]: Oh no! Good catch.

**Comment by colinbrislawn** [comment]
Hey @ajaypanyala, could you review this PR, then Squash and Merge this PR?
Thanks!

**Pull request merged**